javaparser-core/src/main/java/com/github/javaparser/utils/SourceRoot.java

```diff
@@ -92,30 +135,73 @@ public SourceRoot parse(String startPackage, JavaParser javaParser, Callback cal
         assertNotNull(callback);
         Log.info("Parsing package \"%s\"", startPackage);
         final Path path = packageAbsolutePath(root, startPackage);
-        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+        class AsynchronousParse extends RecursiveAction {
+
+            private static final long serialVersionUID = 1L;
+            private final Path path;
+
+            public AsynchronousParse(Path path) {
+                this.path = path;
+            }
+
             @Override
-            public FileVisitResult visitFile(Path absolutePath, BasicFileAttributes attrs) throws IOException {
-                if (!attrs.isDirectory() && absolutePath.toString().endsWith(".java")) {
-                    Path localPath = root.relativize(absolutePath);
-                    Log.trace("Parsing %s", localPath);
-                    final ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(absolutePath));
-                    result.getResult().ifPresent(cu -> cu.setStorage(absolutePath));
-                    if (callback.process(localPath, absolutePath, result) == SAVE) {
-                        if (result.getResult().isPresent()) {
-                            save(result.getResult().get(), path);
+            protected void compute() {
+                final List<AsynchronousParse> walks = new ArrayList<>();
+                try {
+                    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+
+                        @Override
+                        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                                throws IOException {
+                            if (!dir.equals(AsynchronousParse.this.path)) {
+                                AsynchronousParse w = new AsynchronousParse(dir);
+                                w.fork();
+                                walks.add(w);
```

Comment:
I'm actually completely unfamiliar with this API, so I can't really judge if this is well written. Works well for the code generation parts of JP though. @ftomassetti ? @jimshowalter?